lib/blob/blobstore.c

```diff
@@ -7533,6 +7583,38 @@ bs_shallow_copy_blob_read_cpl(void *cb_arg, int bserrno)
 		       &ctx->ext_args);
 }
 
+static void
+bs_range_shallow_copy_cluster_handle_next(void *cb_arg)
+{
+	struct shallow_copy_ctx *ctx = cb_arg;
+	struct spdk_blob *_blob = ctx->blob;
+	struct spdk_bs_dev *ext_dev = ctx->ext_dev;
+
+	if (ctx->range_index < ctx->cluster_count) {
+		ctx->cluster = ctx->clusters_indexes[ctx->range_index];
+
+		if (_blob->active.clusters[ctx->cluster] != 0) {
+			blob_request_submit_op_single(ctx->blob_channel, _blob, ctx->read_buff,
+						      bs_cluster_to_lba(_blob->bs, ctx->cluster),
+						      bs_dev_byte_to_lba(_blob->bs->dev, _blob->bs->cluster_sz),
+						      bs_shallow_copy_blob_read_cpl, ctx, SPDK_BLOB_READ);
+		} else {
+			ctx->ext_args.channel = ctx->ext_channel;
+			ctx->ext_args.cb_fn = bs_shallow_copy_bdev_unmap_cpl;
+			ctx->ext_args.cb_arg = ctx;
+
+			ext_dev->unmap(ext_dev, ctx->ext_channel,
```

Comment:
NIT: Should spdk check if `ext_dev` or the destination bdev supports `unmap`?

Reply:
Wait... Does the below `else` code block mean that spdk_tgt will do nothing but just return 0 (when the destination bdev does not support unmap)?

Reply:
Yes, if the destination bdev doesn't support the unmap, the unmapping is skipped and we only copy the allocated clusters.
Do you need that the operation return error? I could put a check at the beginning of the operation to understand if the lob has unallocated clusters and then check for the unmap supported.

Reply:
I am thinking, if we need to return an error once we can not guarantee the consistency between src and target after this shallow copy.

Reply:
Just inserted a check at the beginning of the operation to understand if the blob has unallocated clusters and then check for the unmap supported.